doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+The functions `train`, `predict`, `explain` ,`run`, `solve`, `evaluate `  provided by each module in `platform` should have the same signature. The signature should be defined in the detailed design doc. **Diagnostics are required**.
+
+###### Golang
+
+To execute a SQLFlow program, the golang package `pkg/executor` should eventually spawn a python process that calls `execute.py:main("execute")`, and passes the filled `Program` to stdin of the process via the standard IPC pipe.
+
+To run a step of a executing SQLFlow program, the golang package `pkg/step` should spawn a python process that calls `execute.py:main("step")`, and passes the filled `Program` (with only one `Statement`) to stdin of the process via the standard IPC pipe.
+
+To get the status of an executing SQLFlow program, the golang package `pkg/server` should eventually spawn a python process that calls `execute.py:main("fetch")` .
+
+###### MVP
+
+The python function `fetch` is **not** required in the MVP.
+
+#### Feature Derivation And Columns
+
+###### Python
+
+The function calls in the COLUMN clause should directly map to calls to python functions in the `columns` package, for example, the `COLUMN NUMERIC(x)` statement would directly map to the python function `numeric()`. We should have at least a `tf_columns.py` module under this package to implement currently supported columns such as `numeric` or `indicator`.
+
+`features.py` should have at least two function: 
+
+```python
+# Columns is the pseudo type of `map<string, Columns>` in the `Statement` message.
+def generate_analysis_steps(datasource:str, columns:Columns):
+    """
+    `generate_analysis_steps` generates a `select` statement to get necessary aggregate data for feature columns
+    """
+    pass
+  
+# ColumnsDAG is a structure to be designed in the detailed design doc.
+def generate_feature_columns(columns: ColumnsDAG):
+    """
+    `generate_feature_columns` generates python feature columns objects
+    """
+    pass
+```
+
+`generate_analysis_steps` does at least the following tasks:
+
+- Call the standard module `ast` to generate ASTs for the function call expressions in `columns`
+- Do semantic checking for every function nodes in the ASTs. **Diagnostics are required.**
+- Collect analysis requirements of the column functions from the nodes in the AST and construct analysis `select` statements
+- Construct a `ColumnsDAG` that represents the `columns` from the AST
+- Return the DAG and the analysis `select` statements to `workflow`. `workflow` should put the DAG in an Argo artifact that could be passed to the following steps. 
+
+`generate_feature_columns` does at least the following task:
+
+- Generate python column objects ( `tf.feature_columns` or a function that returns tensors e.g.) according to the `ColumnsDAG` and the results of the analysis `select` statements (which could be in a database table or in an artifact).
```

Comment:
Generate a python function built upon feature column or keras preprocessing layers. The function accept input features, preprocess them and output tensors.